File: src/main/TransducerSpecification/Transducer/TransducerEdge.java
package Transducer;

// class specifying a transducer edge.
// this edge is more special than the standard query edge.
// we need a 5-tuple, with source and target nodes, incoming and outgoing strings and a cost (doing this replacement operation)
public class TransducerEdge {

    public TransducerNode source;
    public TransducerNode target;
    public String incomingString;
    public String outgoingString;
    public double cost;

    TransducerEdge(TransducerNode source, TransducerNode target, String incomingString, String outgoingString, double cost) {
        this.source = source;
        this.target = target;
        this.incomingString = incomingString.toLowerCase();
        this.outgoingString = outgoingString.toLowerCase();
        this.cost = cost;
    }

    public String toString() {

        String eps = "ε";
        String incoming;
        String outgoing;

        if (incomingString.isEmpty()){
            incoming = eps;
        } else incoming = incomingString;

        if (outgoingString.isEmpty()){
            outgoing = eps;
        } else outgoing = outgoingString;

        return String.format("(%s) -[%3s|%3s|%3s]-> (%s)", source.identifier, incoming, outgoing, cost, target.identifier);
    }

    public void print() {
        System.out.println(toString());
    }
}

File: tinDL/src/main/kotlin/tinDL/model/v1/transducer/TransducerEdge.kt
package tinDL.model.v1.transducer

import tinDL.model.v1.graph.Edge

class TransducerEdge(
    override var source: TransducerNode,
    override var target: TransducerNode,
    var incomingString: String,
    var outgoingString: String,
    var cost: Double
) : Edge(
        source, target, "$incomingString, $outgoingString, $cost"
){

    override fun toString(): String {
        val eps = "epsilon"
        val incoming: String = if (incomingString.isEmpty()) {
            eps
        } else incomingString
        val outgoing: String = if (outgoingString.isEmpty()) {
            eps
        } else outgoingString
        return String.format(
            "(%s) -[%3s|%3s|%3s]-> (%s)",
            source.identifier,
            incoming,
            outgoing,
            cost,
            target.identifier
        )
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is TransducerEdge) return false

        return checkForNodesEquality(other) &&
                incomingString == other.incomingString &&
                outgoingString == other.outgoingString &&
                cost == other.cost
    }
    /**
     * call the trimmed TransducerNode.equals() method in order to prevent a circular dependency.
     */
    private fun checkForNodesEquality(other: TransducerEdge): Boolean {
        return source.equalsWithoutEdges(other.source) &&
                target.equalsWithoutEdges(other.target)
    }

    override fun hashCode(): Int {
        var result = source.identifier.hashCode()
        result = 31 * result + target.identifier.hashCode()
        result = 31 * result + incomingString.hashCode()
        result = 31 * result + outgoingString.hashCode()
        result = 31 * result + cost.hashCode()
        return result
    }
}